File: java/Exercises.java
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Exercises {
    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }

    public static Optional <String> firstThenLowerCase(List<String> strings, Predicate<String> predicate) {
        return strings.stream()
                .filter(predicate)
                .findFirst()
                .map(String::toLowerCase);
    }


    static record Sayer(String phrase){
        Sayer and(String word){
            return new Sayer(phrase + " " + word); 
        }
    }

    public static Sayer say(){
        return new Sayer("");
    }

    public static Sayer say(String word) {
        return new Sayer(word);  
    } 
  

    
    public static int meaningfulLineCount(String filename) throws IOException {
        try (var reader = new BufferedReader(new FileReader(filename))){
            return (int) reader.lines()
                    .filter(line -> !line.trim().isBlank() && !line.trim().startsWith("#"))
                    .count();
        }
    }

}


record Quaternion (double a, double b, double c, double d) {

    public final static Quaternion ZERO = new Quaternion(0, 0, 0, 0);
    public final static Quaternion I = new Quaternion(0, 1, 0, 0);
    public final static Quaternion J = new Quaternion(0, 0, 1, 0);
    public final static Quaternion K = new Quaternion(0, 0, 0, 1);

    public Quaternion {
        if(Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d)) {
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }

    Quaternion plus(Quaternion other) {
        return new Quaternion(a + other.a, b + other.b, c + other.c, d + other.d);
    }

    Quaternion times(Quaternion other) {
        return new Quaternion(
                a * other.a - b * other.b - c * other.c - d * other.d, 
                a * other.b + b * other.a + c * other.d - d * other.c, 
                a * other.c - b * other.d + c * other.a + d * other.b, 
                a * other.d + b * other.c - c * other.b + d * other.a);
    }

    public Quaternion conjugate() {
        return new Quaternion (a, -b, -c, -d);
    }

    public List<Double> coefficients() {
        return List.of(a, b, c, d);
    }

    @Override 
    public String toString() {
        String str = "";
        
        if (a != 0) {
            str += a;
        }
        
        if (b != 0) {
            str += (str.isEmpty() ? "" : (b > 0 ? "+" : "")) + (b == 1 ? "i" : (b == -1 ? "-i" : b + "i"));
        }
    
        if (c != 0) {
            str += (str.isEmpty() ? "" : (c > 0 ? "+" : "")) + (c == 1 ? "j" : (c == -1 ? "-j" : c + "j"));
        }
    
        if (d != 0) {
            str += (str.isEmpty() ? "" : (d > 0 ? "+" : "")) + (d == 1 ? "k" : (d == -1 ? "-k" : d + "k"));
        }
        
        return str.isEmpty() ? "0" : str.replace("+-", "-");
    }
}


sealed interface BinarySearchTree permits Empty, Node {
    int size();
    boolean contains(String value);
    BinarySearchTree insert(String value);
}

final record Empty() implements BinarySearchTree {
    @Override
    public int size() {
        return 0;
    }

    @Override
    public boolean contains(String value) {
        return false;
    }

    @Override
    public BinarySearchTree insert(String value) {
        return new Node(value, this, this);
    }

    @Override
    public String toString() {
        return "()";
    }
}

final class Node implements BinarySearchTree {
    private final String value;
    private final BinarySearchTree left;
    private final BinarySearchTree right;

    Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }

    @Override
    public int size() {
        return 1 + left.size() + right.size();
    }

    @Override
    public boolean contains(String value) {
        return this.value.equals(value) || left.contains(value) || right.contains
        (value);
    }

    @Override
    public BinarySearchTree insert(String value) {
        if (value.compareTo(this.value) < 0) {
            return new Node(this.value, left.insert(value), right);
        } else {
            return new Node(this.value, left, right.insert(value));
        }
    }

    @Override
    public String toString() {
        String leftStr;
        String rightStr;
        if (left instanceof Empty()) {
            leftStr = "";
        } else{
            leftStr = left.toString();
        }

        if (right instanceof Empty()) {
            rightStr = "";
        } else{
            rightStr = right.toString();
        }

        return "(" + leftStr + value + rightStr + ")";
    }

}

File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException


fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}


fun firstThenLowerCase(strings: List<String>, predicate: (String) -> Boolean): String? {
    return strings.firstOrNull(predicate)?.lowercase()
}


data class Say(val phrase: String) {
    fun and(nextPhrase: String): Say {
        return Say("$phrase $nextPhrase")
    }
}

fun say(phrase: String = ""): Say {
    return Say(phrase)
}


@Throws(IOException::class)
fun meaningfulLineCount(filename: String): Long {
    BufferedReader(FileReader(filename)).use { reader ->
        return reader.lines()
        .filter { it.trim().isNotEmpty() }
        .filter { !it.trim().startsWith("#") } 
        .count()
    }
}


data class Quaternion(val a: Double, val b: Double, val c: Double, val d: Double) {
    
    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }
    
    operator fun plus(other: Quaternion): Quaternion {
        return Quaternion(a + other.a, b + other.b, c + other.c, d + other.d)
    }
    
    operator fun times(other: Quaternion): Quaternion {
        return Quaternion(
            a * other.a - b * other.b - c * other.c - d * other.d, 
            a * other.b + b * other.a + c * other.d - d * other.c,
            a * other.c - b * other.d + c * other.a + d * other.b,
            a * other.d + b * other.c - c * other.b + d * other.a
        )
    }

    fun coefficients(): List<Double> = listOf(a, b, c, d)

    fun conjugate(): Quaternion = Quaternion(a, -b, -c, -d)

// used https://kotlinlang.org/docs/control-flow.html#when-expression to help me with the when statement
    override fun toString(): String {
        fun components(num: Double, letter: String): String {
            return when {
                num == -1.0 -> "-$letter" 
                num == 0.0 -> ""
                num == 1.0 -> "+$letter"
                num > 0 -> "+$num$letter"
                else -> "$num$letter"
            }
        }
        
    var result = ""
    if (a != 0.0) {
        result = a.toString()
    }

    var bComponent = components(b, "i")
    var cComponent = components(c, "j")
    var dComponent = components(d, "k")

    result += bComponent + cComponent + dComponent
  
    return if (result.isEmpty()) "0" else result.replace("+-", "-").trimStart('+')
    }
}
    

sealed interface BinarySearchTree {
    fun size(): Int
    fun contains(value: String): Boolean
    fun insert(value: String): BinarySearchTree

    object Empty : BinarySearchTree {
        override fun size(): Int = 0
        override fun contains(value: String): Boolean = false
        override fun insert(value: String): BinarySearchTree = Node(value, Empty, Empty)
        override fun toString(): String = "()"
    }

    data class Node(private val value: String, private val left: BinarySearchTree, private val right: BinarySearchTree) : BinarySearchTree {
        override fun size(): Int = 1 + left.size() + right.size()
        override fun contains(value: String): Boolean = when {
            value < this.value -> left.contains(value)
            value > this.value -> right.contains(value)
            else -> true 
        }

        override fun insert(value: String): BinarySearchTree = when {
            value < this.value -> Node(this.value, left.insert(value), right)
            value > this.value -> Node(this.value, left, right.insert(value))
            else -> this
        }

        override fun toString(): String {
            var leftString = left.toString()
            var rightString = right.toString()
            return "($leftString$value$rightString)".replace("()","")
        }
    } 
    
}  
